File: android/src/main/java/com/margelo/nitro/camera/facedetector/HybridImageFaceDetector.kt
package com.margelo.nitro.camera.facedetector

import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import com.google.mlkit.vision.face.FaceDetection
import com.margelo.nitro.camera.facedetector.extensions.toMLFaceDetectorOptions
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage as MLInputImage
import com.margelo.nitro.NitroModules
import androidx.core.net.toUri

class HybridImageFaceDetector(
  options: ImageFaceDetectorOptions
) : HybridImageFaceDetectorSpec() {
  private val context = NitroModules.applicationContext ?: throw Error("Image Face Detector - No Context available!")
  private val runLandmarks = options.runLandmarks ?: false
  private val runContours = options.runContours ?: false
  private val runClassifications = options.runClassifications ?: false
  private val trackingEnabled = options.trackingEnabled ?: false
  private val faceDetector = FaceDetection.getClient(
    options.toMLFaceDetectorOptions()
  )

  private fun resolveInputImage(
    input: InputImage
  ): String {
    return when (input) {
      is InputImage.First -> input.value
      is InputImage.Third -> input.value.uri
      else -> throw IllegalArgumentException("Invalid image type. Expected string or { uri }")
    }
  }

  private fun createInputImage(uri: String): MLInputImage {
    val parsedUri = uri.toUri()

    return MLInputImage.fromFilePath(
      context,
      parsedUri
    )
  }

  @OptIn(ExperimentalGetImage::class)
  override fun detectFaces(
    image: InputImage
  ): Array<HybridFaceSpec> {
    val uri = resolveInputImage(image)
    val mlImage = createInputImage(uri)
    val config = createFaceProcessConfig(
      frameWidth = mlImage.width.toDouble(),
      frameHeight = mlImage.height.toDouble(),
      autoMode = false,
      runLandmarks = runLandmarks,
      runContours = runContours,
      runClassifications = runClassifications,
      trackingEnabled = trackingEnabled,
      pointTransformer = createIdentityPointTransformer()
    )
    val task = faceDetector.process(mlImage)
    val faces = Tasks.await(task).map {
      HybridFace(it, config)
    }.toTypedArray<HybridFaceSpec>()

    return faces
  }
}

File: ios/HybridImageFaceDetector.swift
import Foundation
import NitroModules
import UIKit
import Vision

class HybridImageFaceDetector: HybridImageFaceDetectorSpec {
  private let runLandmarks: Bool
  private let runContours: Bool
  private let runClassifications: Bool
  private let trackingEnabled: Bool

  init(_ options: ImageFaceDetectorOptions) {
    self.runLandmarks = options.runLandmarks ?? false
    self.runContours = options.runContours ?? false
    self.runClassifications = options.runClassifications ?? false
    self.trackingEnabled = options.trackingEnabled ?? false
    super.init()
  }

  func detectFaces(image: InputImage) throws -> [any HybridFaceSpec] {
    let uiImage = try createInputImage(uri: try resolveInputImage(image))
    guard let cgImage = uiImage.cgImage else {
      throw RuntimeError.error(withMessage: "Failed to create CGImage from input image!")
    }

    let request = VNDetectFaceLandmarksRequest()
    let handler = VNImageRequestHandler(cgImage: cgImage, orientation: .up, options: [:])
    try handler.perform([request])

    let frameWidth = Double(cgImage.width)
    let frameHeight = Double(cgImage.height)
    let config = createFaceProcessConfig(
      frameWidth,
      frameHeight,
      false,
      runLandmarks,
      runContours,
      runClassifications,
      trackingEnabled,
      createIdentityPointTransformer()
    )

    return (request.results ?? []).map { HybridFace(face: $0, config: config) }
  }

  private func resolveInputImage(_ input: InputImage) throws -> String {
    switch input {
      case .first(let uri): return uri
      case .third(let object): return object.uri
      case .second: break
    }
    throw RuntimeError.error(withMessage: "Invalid image type. Expected string or { uri }")
  }

  private func createInputImage(uri: String) throws -> UIImage {
    guard let url = URL(string: uri) else {
      throw RuntimeError.error(withMessage: "Invalid image URI")
    }
    if url.isFileURL, let image = UIImage(contentsOfFile: url.path) { return image }
    if !url.isFileURL, let data = try? Data(contentsOf: url), let image = UIImage(data: data) { return image }
    throw RuntimeError.error(withMessage: "Failed to load image from URI")
  }
}
